### PR TITLE
MinION renovation

### PR DESCRIPTION
**Pull request by kedhammar**
Major overhaul of the way TACA is structured. The intention is to re-make a robust codebase that will facilitate the pending integration of ONT runs to Production for Illumina QC.

The diffs are huge, so the best approach for reviewing may be to checkout the entire PR locally.

# Summary of changes:

## Local script `instrument_transfer.py`
- Harmonized to be used for both ONT instruments.

## TACA method `taca analysis ont-transfer`
- ONT transfer classes are collected in a single file with one parent class and two subclasses for user runs and QC runs. The parent class contains common attributes and methods while the QC run subclass contains new, specific methods.
- TACA analysis entrypoint "ont-transfer" is contained in `nanopore_analysis.py` handles run processing through the sequential logging and execution of class methods, depending on run type.
- Errors are raised on a per-run basis and are sent as a single email containing the full traceback of the error.

## Removing old code
- Unused tests are removed

## Config
- The "nanopore_analysis" section of the taca.yaml config file has been re-structured.
- The following config was used for my local development:

```
nanopore_analysis:

  run_types:

    user_run:
      data_dirs:
        - /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/promethion
        - /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/minion
      ignore_dirs:
        - 'nosync'
        - 'qc'
      instruments:
        promethion:
          transfer_log: /Users/alfred.kedhammar/code/scratch/preproc1/transfer_prom.tsv
          archive_dir: /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/promethion/nosync
          metadata_dir: /Users/alfred.kedhammar/code/scratch/ngi-nas-ns/promethion_data
          destination: ~/promethion
        minion:
          transfer_log: /Users/alfred.kedhammar/code/scratch/preproc1/transfer_minion.tsv
          archive_dir: /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/minion/nosync
          metadata_dir: /Users/alfred.kedhammar/code/scratch/ngi-nas-ns/minion_data
          destination: ~/minion

    qc_run:  
      data_dirs:
        - /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/minion/qc
      ignore_dirs:
        - 'nosync'
      instruments:
        minion:
          transfer_log: /Users/alfred.kedhammar/code/scratch/preproc1/transfer_minion_qc.tsv
          archive_dir: /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/minion/qc/nosync
          metadata_dir: /Users/alfred.kedhammar/code/scratch/ngi-nas-ns/minion_data/qc
          destination: ~/minion/qc
      anglerfish:
        anglerfish_samplesheets_dir: /Users/alfred.kedhammar/code/scratch/ngi-nas-ns/samplesheets/anglerfish/
        anglerfish_env_name: anglerfish
        conda_init_path: /Users/alfred.kedhammar/miniconda3/etc/profile.d/conda.sh

  minknow_reports_dir: /Users/alfred.kedhammar/code/scratch/ngi-internal/minknow_reports
  analysis_server:
    host: preproc1.scilifelab.se  # sätter den till preproc för att det är lättare än miarka för testning
    port:
    user: alfred.kedhammar
  rsync_options:
    #'-Lav': None
    #'--chown': ':ngi2016003'
    #'--chmod' : 'Dg+s,g+rw'
    '-r' : None
    '--exclude' : 'work'
```

### CODE REVIEW

File: taca/analysis/analysis_nanopore.py [hunk 1]
```diff
@@ -1,462 +1,322 @@
...
-    return found_dirs
 
+    logger.info(f"{ont_user_run.run_name}: Touching StatusDB...")
+    ont_user_run.touch_db_entry()
+    logger.info(f"{ont_user_run.run_name}: Touching StatusDB successful...")
+
+    if not ont_user_run.is_synced():
+        logger.info(f"{ont_user_run.run_name}: Run is not fully synced, skipping.")
+    else:
+        logger.info(f"{ont_user_run.run_name}: Run is fully synced, continuning.")
```
Comment (alneberg) [inline taca/analysis/analysis_nanopore.py:78]:
```suggestion
        logger.info(f"{ont_user_run.run_name}: Run is fully synced, continuing.")
```

File: taca/analysis/analysis_nanopore.py [hunk 2]
```diff
@@ -1,462 +1,322 @@
...
+
+    logger.info(f"{ont_qc_run.run_name}: Touching StatusDB...")
+    ont_qc_run.touch_db_entry()
+    logger.info(f"{ont_qc_run.run_name}: Touching StatusDB successful...")
+
+    # Is the run fully synced?
+    if not ont_qc_run.is_synced():
+        logger.info(f"{ont_qc_run.run_name}: Run is not fully synced, skipping.")
+    else:
+        logger.info(f"{ont_qc_run.run_name}: Run is fully synced, continuning.")
```
Comment (alneberg) [inline taca/analysis/analysis_nanopore.py:168]:
```suggestion
        logger.info(f"{ont_qc_run.run_name}: Run is fully synced, continuing.")
```

File: taca/nanopore/ONT_run_classes.py [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+            logger.info(
+                f"{self.run_name}: Successfully created database entry for ongoing run."
+            )
+        else:
+            logger.info(f"{self.run_name}: Database entry already exists, skipping.")
+
+    def update_db_entry(self, force_update=False):
+        """Check run vs statusdb. Create or update run entry."""
+
+        # If no run document exists in the database, ceate an ongoing run document
```
Comment (alneberg) [inline taca/nanopore/ONT_run_classes.py:160]:
```suggestion
        # If no run document exists in the database, create an ongoing run document
```

File: taca/analysis/minion_analysis_WIP.py
Comment (alneberg) [inline taca/analysis/minion_analysis_WIP.py:1]: Was it intentional to bring this file back to life as well with the tests?

File: taca/analysis/analysis_nanopore.py [hunk 4]
```diff
@@ -1,462 +1,322 @@
...
+
+def ont_updatedb(run_abspath: str):
+    """CLI entry function."""
+
+    ont_run = ONT_run(os.path.abspath(run_abspath))
+
+    logger.info(
+        f"{ont_run.run_name}: Manually updating StatusDB, ignoring run status..."
+    )
+    ont_run.update_db(force_update=True)
```
Comment (alneberg) [inline taca/analysis/analysis_nanopore.py:321]:
```suggestion
    ont_run.update_db_entry(force_update=True)
```

**Comment by alneberg** [comment]
Amazing! Need to fix the conflict in versionlog and possibly remove the two minion specific files that was reverted back again? Also, could you have a look at my suggested change if you agree? 

Some tasks for the future:
1. Improve log catching for external commands
2. Think if we want to run more external commands with Popen to enable more parallel execution, but not necessary true that we want this.
3. Get tests functioning again

File: taca/analysis/minion_analysis_WIP.py
Comment (kedhammar) [reply taca/analysis/minion_analysis_WIP.py:1]: [f30934e](https://github.com/SciLifeLab/TACA/pull/386/commits/f30934e489a4ec28a026b0055e1144afa1adfaab)

**Comment by kedhammar** [comment]
New config (simplified Anglerfish call by skipping subprocess Conda and just calling the executable from it's path in the env)

```
nanopore_analysis:
  run_types:
    user_run:
      data_dirs:
        - /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/promethion
        - /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/minion
      ignore_dirs:
        - 'nosync'
        - 'qc'
      instruments:
        promethion:
          transfer_log: /Users/alfred.kedhammar/code/scratch/preproc1/transfer_prom.tsv
          archive_dir: /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/promethion/nosync
          metadata_dir: /Users/alfred.kedhammar/code/scratch/ngi-nas-ns/promethion_data
          destination: ~/promethion
        minion:
          transfer_log: /Users/alfred.kedhammar/code/scratch/preproc1/transfer_minion.tsv
          archive_dir: /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/minion/nosync
          metadata_dir: /Users/alfred.kedhammar/code/scratch/ngi-nas-ns/minion_data
          destination: ~/minion
    qc_run:  
      data_dirs:
        - /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/minion/qc
      ignore_dirs:
        - 'nosync'
      instruments:
        minion:
          transfer_log: /Users/alfred.kedhammar/code/scratch/preproc1/transfer_minion_qc.tsv
          archive_dir: /Users/alfred.kedhammar/code/scratch/preproc1/ngi_data/sequencing/minion/qc/nosync
          metadata_dir: /Users/alfred.kedhammar/code/scratch/ngi-nas-ns/minion_data/qc
          destination: ~/minion/qc
      anglerfish:
        anglerfish_samplesheets_dir: /Users/alfred.kedhammar/code/scratch/ngi-nas-ns/samplesheets/anglerfish/
        anglerfish_path: /Users/alfred.kedhammar/miniconda3/envs/anglerfish/bin/anglerfish
  minknow_reports_dir: /Users/alfred.kedhammar/code/scratch/ngi-internal/minknow_reports
  analysis_server:
    host: preproc1.scilifelab.se  # sätter den till preproc för att det är lättare än miarka för testning
    port:
    user: alfred.kedhammar
  rsync_options:
    #'-Lav': None
    #'--chown': ':ngi2016003'
    #'--chmod' : 'Dg+s,g+rw'
    '-r' : None
    '--exclude' : 'work'
```

**Review by alneberg — APPROVED** [review]
Excellent! No further comments.

File: taca/analysis/analysis_nanopore.py [hunk 5]
```diff
@@ -1,462 +1,295 @@
...
-    except OSError:
-        logger.warning(
-            "There was an issue locating the following directory: {}. "
-            "Please check that it exists and try again.".format(ont_data_dir)
-        )
-    return found_dirs
+def send_error_mail(run_name, error: BaseException):
 
+    email_subject = f"Run processed with errors: {run_name}"
+    email_message = "{}\n\n{}".format(
```
Comment (aanil) [inline taca/analysis/analysis_nanopore.py:41]: Why not use an f-string here too?

**Review by aanil — APPROVED** [review]
👍

File: taca/analysis/analysis_nanopore.py (hunk 5 above)
Comment (kedhammar) [reply taca/analysis/analysis_nanopore.py]: [8f78c86](https://github.com/SciLifeLab/TACA/pull/386/commits/8f78c865c7f28cb945c221f135054a9e8476207e)

**Pull request merged**